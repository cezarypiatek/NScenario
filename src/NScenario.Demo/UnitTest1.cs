using System;
using System.Threading.Tasks;
using NUnit.Framework;

namespace NScenario.Demo
{
    public class Tests
    {

        [Test]
        public async Task should_present_basic_scenario()
        {
            var scenario = TestScenarioFactory.Default();

            await scenario.Step("This is the first step", () =>
            {
                // Here comes the logic
            });
            
            await scenario.Step("This is the second step", () =>
            {
                // Here comes the logic
            });
            
            await scenario.Step("This is the third step", () =>
            {
                // Here comes the logic
            });
        }
        
        

        [Test] 
        public async Task shouldPresentBASICScenarioWithCamelCaseTitle()
        {
            var scenario = TestScenarioFactory.Default();

            await scenario.Step("This is the first step", () =>
            {
                // Here comes the logic
            });
            
            await scenario.Step("This is the second step", () =>
            {
                // Here comes the logic
            });
            
            await scenario.Step("This is the third step", () =>
            {
                // Here comes the logic
            });
        }
        
        [TestCase(false)]
        [TestCase(true)]
        public async Task should_present_basic_scenario_with_explicit_title(bool someFlag)
        {
            var scenario = TestScenarioFactory.Default(title: $"some scenario when flag set to '{someFlag}'");

            await scenario.Step("This is the first step", () =>
            {
                // Here comes the logic
            });
            
            await scenario.Step("This is the second step", () =>
            {
                // Here comes the logic
            });
            
            await scenario.Step("This is the third step", () =>
            {
                // Here comes the logic
            });
        }
        
        [Test]
        public async Task should_present_basic_scenario_with_steps_returning_value()
        {
            var scenario = TestScenarioFactory.Default();

            var valFromStep1 = await scenario.Step("This is the first step", () =>
            {
                // Here comes the logic
                return 1;
            });

            var valFromStep2 = await scenario.Step("This is the second step", async () =>
            {
                // Here comes the logic
                await Task.Yield();
                return valFromStep1 + 1;
            });

            await scenario.Step("This is the third step", () =>
            {
                // Here comes the logic
                _ = valFromStep1 + valFromStep2;
            });
        }
        
        [Test]
        public async Task should_present_scenario_with_sub_steps()
        {
            var scenario = TestScenarioFactory.Default();

            await scenario.Step("This is the first step", async () =>
            {
                await scenario.Step("This is the first sub-step of first step", () =>
                {
                    // Here comes the logic
                });
                await scenario.Step("This is the second sub-step of first step", async () =>
                {
                    await scenario.Step("Yet another nesting level p1", () =>
                    {
                        // Here comes the logic
                    });
                    await scenario.Step("Yet another nesting level p2", () =>
                    {
                        // Here comes the logic
                    });
                });
            });
            
            await PerformReusableScenarioPart(scenario);
            
            await scenario.Step("This is the third step", () =>
            {
                // Here comes the logic
            });
        }

        private static async Task PerformReusableScenarioPart(ITestScenario scenario)
        {
            await scenario.Step("This is the second step", async () =>
            {
                await scenario.Step("This is the first sub-step of second step", () =>
                {
                    // Here comes the logic
                });
                await scenario.Step("This is the second sub-step of second step", () =>
                {
                    // Here comes the logic
                });
            });
        }
    }
}