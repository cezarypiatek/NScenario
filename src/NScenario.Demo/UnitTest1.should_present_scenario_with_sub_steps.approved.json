{
  "ScenarioTitle": "should present scenario with sub steps",
  "MethodName": "should_present_scenario_with_sub_steps",
  "FilePath": "_IGNORED_VALUE_",
  "LineNumber": 106,
  "Status": 0,
  "Steps": [
    {
      "Description": "This is the first step",
      "LineNumber": 108,
      "FilePath": "_IGNORED_VALUE_",
      "ExecutionTime": "_IGNORED_VALUE_",
      "Status": 3,
      "Exception": null,
      "SubSteps": [
        {
          "Description": "This is the first sub-step of first step",
          "LineNumber": 110,
          "FilePath": "_IGNORED_VALUE_",
          "ExecutionTime": "_IGNORED_VALUE_",
          "Status": 3,
          "Exception": null,
          "SubSteps": null
        },
        {
          "Description": "This is the second sub-step of first step",
          "LineNumber": 114,
          "FilePath": "_IGNORED_VALUE_",
          "ExecutionTime": "_IGNORED_VALUE_",
          "Status": 3,
          "Exception": null,
          "SubSteps": [
            {
              "Description": "Yet another nesting level p1",
              "LineNumber": 116,
              "FilePath": "_IGNORED_VALUE_",
              "ExecutionTime": "_IGNORED_VALUE_",
              "Status": 3,
              "Exception": null,
              "SubSteps": null
            },
            {
              "Description": "Yet another nesting level p2",
              "LineNumber": 120,
              "FilePath": "_IGNORED_VALUE_",
              "ExecutionTime": "_IGNORED_VALUE_",
              "Status": 3,
              "Exception": null,
              "SubSteps": null
            }
          ]
        }
      ]
    },
    {
      "Description": "This is the second step",
      "LineNumber": 186,
      "FilePath": "_IGNORED_VALUE_",
      "ExecutionTime": "_IGNORED_VALUE_",
      "Status": 3,
      "Exception": null,
      "SubSteps": [
        {
          "Description": "This is the first sub-step of second step",
          "LineNumber": 188,
          "FilePath": "_IGNORED_VALUE_",
          "ExecutionTime": "_IGNORED_VALUE_",
          "Status": 3,
          "Exception": null,
          "SubSteps": null
        },
        {
          "Description": "This is the second sub-step of second step",
          "LineNumber": 192,
          "FilePath": "_IGNORED_VALUE_",
          "ExecutionTime": "_IGNORED_VALUE_",
          "Status": 3,
          "Exception": null,
          "SubSteps": null
        }
      ]
    },
    {
      "Description": "This is the third step",
      "LineNumber": 129,
      "FilePath": "_IGNORED_VALUE_",
      "ExecutionTime": "_IGNORED_VALUE_",
      "Status": 3,
      "Exception": null,
      "SubSteps": null
    }
  ]
}